File: Stack/20. Valid Parentheses/Solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        # N : len(s)
        # Time Complexity: O(N)
        # Space Complexity: O(N)
        # n = len(s)
        # stack = []
        # for ch in s:
        #     if ch == ')' and len(stack) > 0 and stack[-1] == '(':
        #         stack.pop()
        #     elif ch == '}' and len(stack) > 0 and stack[-1] == '{':
        #         stack.pop()
        #     elif ch == ']' and len(stack) > 0 and stack[-1] == '[':
        #         stack.pop()
        #     else:
        #         stack.append(ch)
        # return True if len(stack) == 0 else False


        # N : len(s)
        # Time Complexity: O(N)
        # Space Complexity: O(N)
        n = len(s)
        bracketMap = {')' : '(', '}' : '{', ']' : '['}
        stack = []
        for ch in s:
            if ch not in bracketMap:
                stack.append(ch)
                continue
            if len(stack) <= 0 or stack[-1] != bracketMap[ch]:
                return False
            stack.pop()
        return True if len(stack) == 0 else False

File: Stack/20. Valid Parentheses/Solution.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        // N : len(s)
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        // int n = s.length();
        // Stack<Character> stack = new Stack<>();
        // for (char ch: s.toCharArray()){
        //     if (ch == ')' && !stack.empty() && stack.peek() == '(') stack.pop();
        //     else if (ch == '}' && !stack.empty() && stack.peek() == '{') stack.pop();
        //     else if (ch == ']' && !stack.empty() && stack.peek() == '[') stack.pop();
        //     else stack.push(ch);
        // }
        // return stack.empty() ? true : false;


        // N : len(s)
        // Time Complexity: O(N)
        // Space Complexity: O(N)
        int n = s.length();
        HashMap<Character, Character> bracketMap = new HashMap<>();
        bracketMap.put(')', '(');
        bracketMap.put('}', '{');
        bracketMap.put(']', '[');
        Stack<Character> stack = new Stack<>();
        for (char ch: s.toCharArray()){
            if (!bracketMap.containsKey(ch)){
                stack.push(ch);
                continue;
            }
            if (stack.empty() || stack.peek() != bracketMap.get(ch)) return false;
            stack.pop();
        }
        return stack.empty() ? true : false;
    }
}